### DISCUSSION TITLE
conditions for no periodic solutions

### DISCUSSION THREAD

**Discussion by irgallard** [ocelot-collab/ocelot#175] [General]
Hello,

I am trying to create a DRL environment with Ocelot, with this default configuration of elements in the lattice:  https://nbviewer.org/github/ocelot-collab/ocelot/blob/master/demos/ipython_tutorials/1_introduction.ipynb

To achieve this, I have been analyzing how the Twiss parameters change with small increments in the strength of the quadrupoles. However, during my experiments, I encountered a particular issue with some configurations and changes that resulted in the following situation:

![Screenshot from 2023-07-26 12-24-50](https://github.com/ocelot-collab/ocelot/assets/115991988/4d56cfda-de65-4b2f-889e-8a414785a899)

Are there any guidelines to follow in order to update the strengths? Managing multiple magnets at once maybe? 
If so, it would be nice to know how to avoid such solutions. Otherwise I am considering their use as a  "game over" condition for the learning algorithm.

Thank you all for your time and support!

Best regards,
Galo

**Comment by sergey-tomin** [comment]
Hi,

do you need  a periodic solution for your lattice every time ? If yes, you can look in a accelerator handbook the condition for existence of the periodic solution.   Or you can look in ocelot: lines  152-157 https://github.com/ocelot-collab/ocelot/blob/master/ocelot/cpbd/optics.py 
maybe you can check this condition earlier

**Comment by st-walker** [comment]
As Sergey says, some lattices are fundamentally unstable and have no stable solutions.  Often this applies to when quadrupoles become too strong and over-focussing (as opposed to too weak).  

How to adjust quadrupoles?

If you have a `MagneticLattice` instance:

```
mlat = MagneticLattice(my_sequence)
quads_i_want = ["quad1", "quad2", "quad4"]
quads = {element.id: element for element in mlat.sequence if element.id in quads_i_want}
quads["quad1"].k1 = 3.0
quads["quad2"].k1= 1
# etc...
```

Actually I wish we had a native container type that allowed for addressing elements within a sequence easily by name.  That would make this all much easier.  Currently we just use lists.

**Reply by irgallard** [reply]
Thanks  for both of your replies!

I see now the condition in the code @sergey-tomin  shared, basically cosines in both coordinates must exist. I have been looking for it in the textbooks but I havent found it yet. 

Regarding the magnets adjustmentes  @st-walker mentioned, yes, I use something like that in my code aswell with small increments of k for each quadrupole.  I was wondering if the condition of the previous paragraph translated directly into a way of changing the magnets physically with k (besides bounding their strength to certain range), but I guess we have to check it with the matrix R every time a step is taken, right?

**Reply by st-walker** [reply]
https://www-eng.lbl.gov/~dleitner/USPAS_2016_Fundamental_Of_Ion_Sources/Reference%20Material/ParticleAcceleratorPhysics_HelmutWiedemann.pdf

page 244 of this book has what you need to know for stability of whole lattice.

previous section of the book also has formulae for fodo cell stability 

i don't know which makes more sense for you because i don't know anything about your machine

**Reply by irgallard** [reply]
Thanks for the reference! 

So the machine is going to be a deuterons linear accelerator called IFMIF-DONES:
![image](https://github.com/ocelot-collab/ocelot/assets/115991988/614e845b-5cee-4ed2-b8f6-6f28add2415e)

The idea of the work is to research the potential of DRL applied to autonomous control of the beam, since the project is just in its early stages. For now, It can be simple lattices and not necessarily representative of the final accelerator, as long as the configurations are physically correct .  And with it, we want to train an angent that can steer, focus or change the energy of the beam modifying the actuator magnets in small steps.

Some works have done things like this already, one of them is for ARES at DESY (they use Ocelot too, you may know the authors), with simple setups:
https://publikationen.bibliothek.kit.edu/1000141966
![image](https://github.com/ocelot-collab/ocelot/assets/115991988/6ad3ec01-257e-435c-b1c8-1ef1d5b5eb15)

**Reply by st-walker** [reply]
OK in which case I am not sure how much it makes sense to even consider stability, unless you are strictly thinking of FODO cell stability, because your machine is a linac.  In which case it's possible you are using the `twiss` function wrong.  If you provide no tws0 kwarg then it will look for a stable solution.  If you provide a `tws0` kwarg then it will simply track from the start of the lattice these twiss parameters until the end (no consideration of stable solution).  It's true that this is not so clear or obvious.  Ideally we would just have two functions: one for finding periodic twiss and another for tracking twiss in a straight line, then it would be much clearer.

**Reply by irgallard** [reply]
#### `ocelot.twiss(lattice, tws0=None, nPoints=None)`
- `tws0`: initial twiss parameters, `Twiss()` object. If `None`, try to find periodic solution.

Based on your reply and the documentation (above),  then I can just run the simulations for a particle array without trying to find a stable solution (periodic), for any magnetic lattice.

Would you say that not all linear accelerators need to  have stable solutions, depending on the accelerator's objective? (i.e delivering a beam with certain shape and energy to a target).

**Reply by st-walker** [reply]
>Based on your reply and the documentation (above), then I can just run the simulations for a particle array without trying to find a stable solution (periodic), for any magnetic lattice.

Well not a ParticleArray necessarily but some initial tws0 `Twiss` instance.  Then it will just propagate the twiss parameters to the end of the lattice.

>Would you say that not all linear accelerators need to have stable solutions, depending on the accelerator's objective? (i.e delivering a beam with certain shape and energy to a target).

Yes.  A linac having no stable solution means nothing.  It doesn't need to be stable because it's single pass.

**Reply by irgallard** [reply]
Thanks! My last two questions now:

- Following up what you just said, how do we know the initial Twiss parameters we have to input in each case/configuration?
- When you inject particles for the particle arrray you can track the beam through the lattice, and even check the final distribution of particles for this specific set up (like here, 20k: [Notebook](https://nbviewer.org/github/ocelot-collab/ocelot/blob/master/demos/ipython_tutorials/accelerator_optim.ipynb)).  In what amount of time does this process happen? What is the injection rate?

**Reply by st-walker** [reply]
>Following up what you just said, how do we know the initial Twiss parameters we have to input in each case/configuration?

Well that's for you to know.  It depends on *your* machine.  Ultimately it depends on how you generated your charged particle beam and how you manipulated it (in real life) with electromagnetic fields.  We have a photocathode here, then a solenoid and accelerating cavities.  LHC uses a bottle of hydrogen and then I don't know what.  I don't know what you use.

>When you inject particles for the particle arrray you can track the beam through the lattice, and even check the final distribution of particles for this specific set up (like here, 20k: [Notebook](https://nbviewer.org/github/ocelot-collab/ocelot/blob/master/demos/ipython_tutorials/accelerator_optim.ipynb)). In what amount of time does this process happen? 

The main computation goes into the collective effects calculations, not the magnetic tracking.  So space charge, coherent synchrotron radiation effects, these take time.  But maybe for your beam if it is at a high energy then you maybe don't care about space charge, and if the bunch is long you don't care about coherent synchrotron radiation effects, and so on.  If you do not care about collective effects then tracking 20k particles is very fast, but how long is a piece of string?  How fast depends on how long your machine is.  But for EuXFEL injector to track 50m without any collective effects it is on the order of several seconds.  If you want to know how long that jupyter lab example takes to run, then you should run it and time it.

>What is the injection rate?


I don't know what this means.

**Reply by irgallard** [reply]
I see! 

With injection rate I meant how fast these 20k particles are introduced into the lattice.  All at once?

**Reply by st-walker** [reply]
Yes the particles are tracked all at once.  In general this is more computationally efficient.  If you read about CPU caches and vectorisation  then you will understand why this is.

**Comment by sergey-tomin** [comment]
you can address by element itself. See example:
```
q1 = Quadrupole(l=0.5, k1=1)
q2 = Quadrupole(l=0.5, k1=-1)
d = Drift(l=1)
cell = [q1, d, q2]
lat = MagneticLattice(cell)

# you can change quads strength in any moment 
q1.k1  = 2
q2.k1 = -2
```


etc.
